### PR TITLE
A function for applying stochastic gradient descent to a Liesel model

### PR DESCRIPTION
**Pull request by jobrachem**
This PR introduces the function `gs.optim`.

This function provides a quick-and-easy way to apply stochastic gradient descent to a `lsl.Model` - which can be nice for providing a "warm start" for MCMC sampling. All loops and batching are implemented in JAX.

Optimization can be applied on the whole dataset or in batches. By default, the Adam optimizer is used with a fixed learning rate of `0.1`. Users can supply an optimizer of their choice from the `optax` library.

The PR includes documentation and unit tests.

Here's a notebook for playing around:
[051-optim.ipynb.zip](https://github.com/liesel-devs/liesel/files/13780492/051-optim.ipynb.zip)

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -31,6 +31,7 @@ install_requires =
     seaborn>=0.13
     tensorflow-probability>=0.17
     tqdm>=4.62
+    optax
```
Comment (wiep) [inline setup.cfg:34]: I think it needs a version requirement like the other dependencies.

**Comment by jobrachem** [comment]
We should think about the batching and how/whether it generalizes to different models. right now it assumes iid data.

**Comment by jobrachem** [comment]
paul: maybe give it a different name that clearly identifies that this assumes iid data

**Comment by jobrachem** [comment]
I had a quick look at the code again. This is the function that does the batching. It is applied to `Var` objects that are marked as `Var.observed`. To be honest, I don't see yet how this implies an assumption of iid data. 

```python
def batched_nodes(nodes: dict[str, Array], batch_indices: Array) -> dict[str, Array]:
     """Returns a subset of the graph state using the given batch indices."""
     return jax.tree_util.tree_map(lambda x: x[batch_indices, ...], nodes)
```

**Comment by wiep** [comment]
It assumes that each node marked as observed can be indexed in the same way. 

Consider a hierarchical model (and do not worry about the validity of the estimation):

Let's assume that $Y_{ij}$ is the outcome variable for the $i$-th level-1 unit in the $j$-th level-2 unit. The covariates at level-1 and level-2 are represented by $X_{ij}$ and $W_{j}$, respectively. The model can be represented as follows:

Level-1 model (within groups):
$$Y_{ij} = \beta_{0j} + \beta_{1j}X_{ij} + \epsilon_{ij}$$
where $\epsilon_{ij}$ is the level-1 error term.

Level-2 model (between groups):
$$\beta_{0j} = \gamma_{00} + \gamma_{01}W_{j} + u_{0j}$$
$$\beta_{1j} = \gamma_{10} + \gamma_{11}W_{j} + u_{1j}$$
where $u_{0j}$ and $u_{1j}$ are level-2 error terms.

The design matrices $\boldsymbol{W}$ and $\boldsymbol{X}$, arising from stacking the covariates, have different number of rows and cannot be indexed in the same way.

**Comment by jobrachem** [comment]
Thanks for the clarification @wiep !

> It assumes that each node marked as observed can be indexed in the same way.

Indeed it does. I would call this a "flat" way of indexing, as opposed to "nested" indexing. Pointing this out in a way that cannot be missed or is at least hard to miss would be important. This limitation only applies to batching though, the rest of the function does not make such an assumption. So I am not certain whether changing the name of the function to something like `optim_flat` would be necessary - it seems a bit overly restrictive to me.

Maybe the function can be designed in a way that allows the use of different, also user-supplied batching strategies. I would have to take a closer look at the code again to see whether this may even be easy to implement.


> Consider a hierarchical model (and do not worry about the validity of the estimation):
> 
> Let's assume that Yij is the outcome variable for the i-th level-1 unit in the j-th level-2 unit. The covariates at level-1 and level-2 are represented by Xij and Wj, respectively. The model can be represented as follows:
> 
> Level-1 model (within groups): Yij=β0j+β1jXij+ϵij where ϵij is the level-1 error term.
> 
> Level-2 model (between groups): β0j=γ00+γ01Wj+u0j β1j=γ10+γ11Wj+u1j where u0j and u1j are level-2 error terms.
> 
> The design matrices W and X, arising from stacking the covariates, have different number of rows and cannot be indexed in the same way.

I do see the point that this model, like it is, would not fit into the batching scheme. You can, however, write down the model in a "flat" way that allows the batching to work. Still, of course that does not eliminate the fact that the `batch_size` argument of the proposed `optim` function has this limitation. This makes it even clearer how important it would be to alert users to the way batching is done.

**Comment by GianmarcoCallegher** [comment]
I think that having different batching strategies would be really nice. Consider also the case in which you have a binary response and a really unbalanced dataset with a lot of 0s and only a few 1s. You might want a batching strategy that guarantees to always have some 1s in each batch

**Comment by jobrachem** [comment]
We will use `optim_flat` for now, this also gives us room to favor a different implementation under the name `optim` later without backwards compatibility issues.

**Comment by jobrachem** [comment]
@wiep if you got a notification for this, you can ignore it. @GianmarcoCallegher is doing the review.

File: liesel/goose/optim.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+    if batch_size is None:
+        batch_size = n
+
+    n_batches_inside = n // batch_size
+    batch_indices = jnp.arange(0, batch_size)
+    last_batch_size = n % batch_size
+    last_batch_start = n_batches_inside * batch_size
+    last_batch_end = last_batch_start + last_batch_size
+    last_batch_indices = jnp.arange(last_batch_start, last_batch_end)
```
Comment (GianmarcoCallegher) [inline liesel/goose/optim.py:180]: I think it would be better to change the batch indexes at every iteration. This usually helps to avoid local minima and improves generalization. Also, you can have the last batch of the same size as the others by resampling some data points without replacement.

```python
def _generate_batches_idx(slef, key: JaxArray, N: int, batch_size: int) -> JaxArray:
        key_permutation, key_choice = random.split(key, 2)
        batches_idx = jnp.hstack([
            jax.random.permutation(key_permutation, N),
            jax.random.choice(key_choice, N, (batch_size - N % batch_size, ))
        ])
        return jnp.array_split(batches_idx, len(batches_idx) // batch_size)
```

File: liesel/goose/optim.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        pos = optax.apply_updates(pos, updates)
+        return pos, opt_state
+
+    def _update_cond(val):
+        i, _, _, nlp_old, nlp_new = val
+
+        abs_change = jnp.abs(nlp_old - nlp_new)
+        rel_change = jnp.abs(abs_change / nlp_old)
+
+        return (abs_change > atol) & (rel_change > rtol) & (i < maxiter)
```
Comment (GianmarcoCallegher) [inline liesel/goose/optim.py:204]: Here we are checking the absolute change and the relative change between epoch $t - 1$ and $t$. I think it would be better to introduce a patience of more epochs. It might be that in some cases we are not monotonically decreasing

**Comment by GianmarcoCallegher** [comment]
Do we want to always have the negative log probability as the function to optimize or do we want to provide another parameter which could be the function to optimize, (e.g. RMSE)?

File: liesel/goose/optim.py (hunk 2 above)
Comment (jobrachem) [reply liesel/goose/optim.py:180]: Oh, so this would mean you don't split up the data into batches of a certain size, but you always draw a random sample of the batch size with replacement? So the same data points can get used twice?

File: liesel/goose/optim.py (hunk 3 above)
Comment (jobrachem) [reply liesel/goose/optim.py:204]: That's a very good point. So far, I did not implement patience, because the algorithm runs without creating a history. But probably it should do that, and then it would be possible to implement some patience.

File: liesel/goose/optim.py (hunk 2 above)
Comment (GianmarcoCallegher) [reply liesel/goose/optim.py:180]: You split the data in batches of a certain size but you change the elements in each batch every iteration. It might be that N % batch_size != 0. In that case, you can resample some data points to fill the batch. It is like doing a "data augmentation".

Example:

You have 5 data points (x_1, x_2, x_3, x_4, X_5) and a batch size of 2:

Iteration 1: (x_1, x_3), (x_4, x_2), (x_5, x_3) sampled randomly
Iteration 2: (x_4, x_5), (x_2, x_1), (x_3, x_4) sampled randomly
Iteration3: ...

Comment (jobrachem) [reply liesel/goose/optim.py:180]: That sounds good! Thanks a lot! 😊

**Comment by jobrachem** [comment]
> Do we want to always have the negative log probability as the function to optimize or do we want to provide another parameter which could be the function to optimize, (e.g. RMSE)?

Also a really  good point. I guess it would be relatively easy to let users provide their own function for optimization.

File: liesel/goose/optim.py (hunk 2 above)
Comment (jobrachem) [reply liesel/goose/optim.py:180]: @wiep suggests dropping the last, non-full batch instead of filling it via resampling. This is probably more common.

**Review by GianmarcoCallegher — CHANGES_REQUESTED** [review]
You are a good boy. Do the changes ❤️

File: liesel/goose/optim.py (hunk 2 above)
Comment (GianmarcoCallegher) [reply liesel/goose/optim.py:180]: I haven't found a paper where they prove that adding some resample observations to the last batch leads to an unbiased estimator, but here are two example where they do that:

- [neon](https://neon.nervanasys.com/docs/latest/faq.html)
- [CNN Sentence](https://github.com/yoonkim/CNN_sentence/blob/master/conv_net_sentence.py) from this popular paper [Convolutional Neural Networks for Sentence Classification](https://arxiv.org/pdf/1408.5882.pdf%EF%BC%89%E3%80%82%E4%B8%8B%E9%9D%A2%E7%9C%8B%E4%B8%8B%E8%AE%BA%E6%96%87%E4%B8%AD%E7%9A%84%E6%88%AA%E5%9B%BE%EF%BC%9A)

Comment (jobrachem) [reply liesel/goose/optim.py:180]: Thanks for the research! :)

**Comment by jobrachem** [comment]
Okay, I finally implemented a big update. Now we have the following:

1. Batching is implemented as suggested by @GianmarcoCallegher (i.e. randomly assembled batches in each iteration) with the modification suggested by @wiep (i.e. if the number of observations is not divisible by the batch size, the last batch is simply dropped in that iteration).
2. The function now saves a history of training and (if applicable) testing loss, as well as a history of parameter values. The latter can be turned off to save memory.
3. Early stopping is now implement with patience.
4. If the users wishes, they can supply a `test_model`. If supplied, the test loss is used to determine early stopping. For fully Bayesian optimization I think using a test model usually be a little weird, because it affects the balance of likelihood and prior. But there might be cases in which one would want to fit a likelihood model with fully constant priors.

Here's a notebook with a minimal example for testing:

[075-optim.ipynb.zip](https://github.com/liesel-devs/liesel/files/14733605/075-optim.ipynb.zip)

Do you think the current defaults are good?
1. Especially the defaults for the `stopper` are quite arbitrarily chosen right now. 
2. `save_position_history=True`
3. `restore_best_position=True`
4. `prune_history=True`

**Comment by GianmarcoCallegher** [comment]
Thank you very much, @jobrachem. 

Some general remarks:
- I would replace the word "test" with the word "validation"
- Usually, early stopping is performed only if there is a validation dataset which is different from the training dataset, but here, if we don't have the validation dataset, we set it to be equal to the train one `model_test = model_test if model_test is not None else model`. If there is no validation set, you just run the model for all epochs on the training set

What do you think? ❤️

**Comment by jobrachem** [comment]
Thank you very much for the speedy review @GianmarcoCallegher ! Your suggestions make sense to me, I will implement them.

**Comment by GianmarcoCallegher** [comment]
> Okay, I finally implemented a big update. Now we have the following:
> 
> 1. Batching is implemented as suggested by @GianmarcoCallegher (i.e. randomly assembled batches in each iteration) with the modification suggested by @wiep (i.e. if the number of observations is not divisible by the batch size, the last batch is simply dropped in that iteration).
> 2. The function now saves a history of training and (if applicable) testing loss, as well as a history of parameter values. The latter can be turned off to save memory.
> 3. Early stopping is now implement with patience.
> 4. If the users wishes, they can supply a `test_model`. If supplied, the test loss is used to determine early stopping. For fully Bayesian optimization I think using a test model usually be a little weird, because it affects the balance of likelihood and prior. But there might be cases in which one would want to fit a likelihood model with fully constant priors.
> 
> Here's a notebook with a minimal example for testing:
> 
> [075-optim.ipynb.zip](https://github.com/liesel-devs/liesel/files/14733605/075-optim.ipynb.zip)
> 
> Do you think the current defaults are good?
> 
> 1. Especially the defaults for the `stopper` are quite arbitrarily chosen right now.
> 2. `save_position_history=True`
> 3. `restore_best_position=True`
> 4. `prune_history=True`

Concerning the default values, I think they are totally fine. Maybe I would not have a default number of iterations, which has to be specified by the user

**Comment by jobrachem** [comment]
Over lunch, @wiep mentioned the practice of rescaling the likelihood when using mini-batching in order to ensure that the prior does not get more power than warranted. I think this should be implemented in our function here, too. It can be used both in general for minibatching, and also in order to correct for a potentially smaller sample size in a validation model.

File: liesel/goose/optim.py [hunk 4]
```diff
@@ -277,13 +297,17 @@ def optim_flat(
     batch_seed = (
         batch_seed if batch_seed is not None else np.random.randint(low=1, high=1000)
     )
-    model_test = model_test if model_test is not None else model
+
+    user_patience = stopper.patience
+    if model_validation is None:
+        model_validation = model_validation if model_validation is not None else model
```
Comment (GianmarcoCallegher) [inline liesel/goose/optim.py:303]: If `model_validation` is None, you can already assign `model_validation = model` without the inline if statement, right?

**Review by GianmarcoCallegher — CHANGES_REQUESTED** [review]
Thanks a lot, @jobrachem . Everything looks great. I left a couple of more comments

File: liesel/goose/optim.py (hunk 4 above)
Comment (jobrachem) [reply liesel/goose/optim.py:321]: Good point! 😅

File: liesel/goose/optim.py [hunk 5]
```diff
@@ -106,34 +106,54 @@ class Stopper:
     max_iter: int
     patience: int
     atol: float = 1e-3
-    rtol: float = 1e-6
+    rtol: float = 1e-12
```
[the remark this answers is not in the document]
Comment (jobrachem) [reply liesel/goose/optim.py:110]: It is very strict - with this default, early stopping basically only depends on the absolute tolerance. That is intentional here for the default.

**Comment by jobrachem** [comment]
@GianmarcoCallegher I implemented my changes. If you are satisfied, please squash & merge 😊

File: liesel/goose/optim.py [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+    )
+    _validate_log_prob_decomposition(
+        interface_validation, position=position, state=model_validation.state
+    )
+
+    # ---------------------------------------------------------------------------------
+    # Define loss function(s)
+
+    likelihood_scalar = n_train / batch_size
+    likelihood_scalar_validation = n_train / n_validation
```
Comment (GianmarcoCallegher) [inline liesel/goose/optim.py:356]: I'm just curious. Why is that?

**Comment by GianmarcoCallegher** [comment]
Everything LGTM, @jobrachem . I just left one final comment (probably it is not an error, I am just confused)

File: liesel/goose/optim.py (hunk 6 above)
Comment (jobrachem) [reply liesel/goose/optim.py:356]: Thanks for the question @GianmarcoCallegher !
Ah the idea here is to rescale the likelihood to keep its ratio to the prior constant, as discussed here: https://github.com/liesel-devs/liesel/pull/175#issuecomment-2030409019

Otherwise, if you use batching or a small test dataset, the prior gets relatively more important than if you were to use the likelihood. 

Does that make sense in your opinion?

**Comment by jobrachem** [comment]
Thanks @GianmarcoCallegher ! See my reply above :)
